Multisig/Data/Services/Notification Service/RegisterNotificationTokenRequest.swift

```diff
@@ -7,27 +7,58 @@
 //
 
 import Foundation
+import Web3
 
 struct RegisterNotificationTokenRequest: JSONRequest {
-    let uuid: String?
+    let uuid: String
     let safes: [String]
     let cloudMessagingToken: String
     let bundle: String
     let version: String
     let deviceType: String = "IOS"
     let buildNumber: String
+    let timestamp: String
+    let signatures: [String]?
+
     var httpMethod: String { return "POST" }
     var urlPath: String { return "/api/v1/notifications/devices/" }
 
     typealias ResponseType = Response
 
-    init(deviceID: UUID? = nil, safes: [Address], token: String, bundle: String, version: String, buildNumber: String) {
-        self.uuid = deviceID?.uuidString.lowercased()
+    init(deviceID: UUID,
+         safes: [Address],
+         token: String,
+         bundle: String,
+         version: String,
+         buildNumber: String,
+         timestamp: String = String(Int(Date().timeIntervalSince1970 * 1_000))) throws {
+
+        self.uuid = deviceID.uuidString.lowercased()
         self.safes = safes.map { $0.checksummed }
         self.cloudMessagingToken = token
         self.bundle = bundle
         self.version = version
         self.buildNumber = buildNumber
+        self.timestamp = timestamp
+
+        let string = [
+            "gnosis-safe",
+            self.uuid,
+            self.safes.reduce("") { $0 + $1 },
+            self.cloudMessagingToken,
+            self.bundle,
+            self.version,
+            self.deviceType,
+            self.buildNumber,
+            self.timestamp
+        ]
+        .reduce("") { $0 + $1 }
+
+        if let signature = try? Signer.sign(string).value {
+            self.signatures = [signature]
+        } else {
+            self.signatures = nil
```

Comment:
if there are no signatures, then `timestamp` must be omitted

Reply:
Ah, that what you mean under "optional"? I think it is not required to do so for proper backend calculations, because it will just not participate in the logic. There is no such thing a strict rule not to send some optional params for backends, but there is a rule to always send them if backend expects them to have.

Reply:
this is how it was agreed upon in the spec, we'd better not assume backend implementation details when we implement our code

Reply:
ok